certificates/src/lib.rs

```diff
@@ -21,30 +23,60 @@ pub enum CertificateError {
     CertificateExpired,
 
     #[error("Failed to parse private key")]
-    BadPrivateKey(openssl::error::ErrorStack),
+    BadPrivateKey(#[source] openssl::error::ErrorStack),
 
     #[error("Certificate and private key do not match")]
     Mismatch,
 
-    #[error("Unexpected error: {0}")]
-    Unexpected(openssl::error::ErrorStack),
+    #[error("Failed to convert common name to UTF8")]
+    NonUtf8CommonName(#[source] openssl::error::ErrorStack),
+
+    #[error("Certificate subject does not contain a common name")]
+    NoCommonName,
+
+    #[error("Certificate contains no subject alternate name dNSName entries and multiple common names")]
+    NoDnsNameSansMultipleCommonNames,
```

Comment:
```suggestion
    AmbiguousName,
```

(feel free to ignore.  I misread `Sans` to mean "without" and was looking for something less of a mouthful to put here.)